client/java/src/main/java/com/linecorp/centraldogma/client/AbstractCentralDogmaBuilder.java

```diff
@@ -214,53 +224,65 @@ public final B profile(ClassLoader classLoader, Iterable<String> profiles) {
+        try {
+            final List<ClientProfile> availableProfileList =
+                    new ObjectMapper().readValue(resourceUrl, new TypeReference<List<ClientProfile>>() {});
+            availableProfiles = availableProfileList.stream()
+                                                    .collect(toImmutableMap(ClientProfile::name,
+                                                                            Function.identity()));
+        } catch (IOException e) {
+            throw new IllegalStateException("failed to load: " + PROFILE_RESOURCE_PATH, e);
+        }
+
+        final List<String> reversedProfiles = reverse(profiles);
+        checkArgument(!reversedProfiles.isEmpty(), "profiles is empty.");
+        for (String candidateName : reversedProfiles) {
+            checkNotNull(candidateName, "profiles contains null: %s", profiles);
 
-            final String path = "centraldogma-profile-" + p + ".properties";
-            final InputStream in = classLoader.getResourceAsStream(path);
-            if (in == null) {
+            final ClientProfile candidate = availableProfiles.get(candidateName);
+            if (candidate == null) {
                 continue;
             }
 
-            try {
-                final ImmutableSet.Builder<InetSocketAddress> newHostsBuilder = ImmutableSet.builder();
-                final Properties props = new Properties();
-                props.load(in);
-                for (Entry<Object, Object> e : props.entrySet()) {
-                    final String key = (String) e.getKey();
-                    final String value = (String) e.getValue();
-
-                    if (key.startsWith("centraldogma.hosts.")) {
-                        final HostAndPort hostAndPort = HostAndPort.fromString(value);
-                        final InetSocketAddress addr = newEndpoint(
-                                hostAndPort.getHost(), hostAndPort.getPortOrDefault(DEFAULT_PORT));
-                        newHostsBuilder.add(addr);
-                    }
-                }
-
-                final ImmutableSet<InetSocketAddress> newHosts = newHostsBuilder.build();
-                if (newHosts.isEmpty()) {
-                    printWarning(String.format("Ignoring %s: contains no hosts", path));
-                } else {
-                    selectedProfile = p;
-                    hosts = newHosts;
-                }
-            } catch (IOException e) {
-                throw new IllegalArgumentException("failed to load: " + path, e);
+            final ImmutableSet.Builder<InetSocketAddress> newHostsBuilder = ImmutableSet.builder();
+            candidate.hosts().stream()
+                     .filter(e -> (useTls ? "https" : "http").equals(e.protocol()))
+                     .forEach(e -> newHostsBuilder.add(newEndpoint(e.host(), e.port())));
+
+            final ImmutableSet<InetSocketAddress> newHosts = newHostsBuilder.build();
+            if (!newHosts.isEmpty()) {
```

Comment:
How about adding a warning log message like what we did before when `newHosts` is empty?

Reply:
I'm not sure if it's a good idea because it can be triggered even when `hosts` is not empty but there's no matching protocol.

Reply:
I misread the previous code. Thanks! :-)